dbms/src/DataTypes/DataTypeDateTime64.cpp

```diff
@@ -0,0 +1,257 @@
+    value_index = static_cast<bool>(protobuf.writeDateTime64(assert_cast<const ColumnType &>(column).getData()[row_num], scale));
+}
+
+void DataTypeDateTime64::deserializeProtobuf(IColumn & column, ProtobufReader & protobuf, bool allow_add_row, bool & row_added) const
+{
+    row_added = false;
+    DateTime64 t = 0;
+    if (!protobuf.readDateTime64(t, scale))
+        return;
+
+    auto & container = assert_cast<ColumnType &>(column).getData();
+    if (allow_add_row)
+    {
+        container.emplace_back(t);
+        row_added = true;
+    }
+    else
+        container.back() = t;
+}
+
+bool DataTypeDateTime64::equals(const IDataType & rhs) const
+{
+    if (auto * ptype = typeid_cast<const DataTypeDateTime64 *>(&rhs))
+        return this->scale == ptype->getScale();
+    return false;
+}
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+enum class ArgumentKind
+{
+    Optional,
+    Mandatory
+};
+
+template <typename T, ArgumentKind Kind>
+std::conditional_t<Kind == ArgumentKind::Optional, std::optional<T>, T>
+getArgument(const ASTPtr & arguments, size_t argument_index, const char * argument_name, const std::string context_data_type_name)
+{
+    using NearestResultType = NearestFieldType<T>;
+    const auto fieldType = Field::TypeToEnum<NearestResultType>::value;
+    const ASTLiteral * argument = nullptr;
+
+    auto exceptionMessage = [=](const String & message)
+    {
+        return std::string("Parameter #") + std::to_string(argument_index) + " '"
+                + argument_name + "' for " + context_data_type_name
+                + message
+                + ", expected: " + Field::Types::toString(fieldType) + " literal.";
+    };
+
+    if (!arguments || arguments->children.size() <= argument_index
+        || !(argument = arguments->children[argument_index]->as<ASTLiteral>()))
+    {
+        if constexpr (Kind == ArgumentKind::Optional)
+            return {};
+        else
+            throw Exception(exceptionMessage(" is missing"),
+                            ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+    }
+
+    if (argument->value.getType() != fieldType)
+        throw Exception(exceptionMessage(String(" has wrong type: ") + argument->value.getTypeName()),
+                        ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+    return argument->value.get<NearestResultType>();
+}
+
+static DataTypePtr create64(const ASTPtr & arguments)
+{
+    if (!arguments || arguments->size() == 0)
+        return std::make_shared<DataTypeDateTime64>(DataTypeDateTime64::default_scale);
+
+    const auto scale = getArgument<UInt64, ArgumentKind::Optional>(arguments, 0, "scale", "DateType64");
+    const auto timezone = getArgument<String, ArgumentKind::Optional>(arguments, !!scale, "timezone", "DateType64");
+
+    return std::make_shared<DataTypeDateTime64>(scale.value_or(DataTypeDateTime64::default_scale), timezone.value_or(String{}));
```

Comment:
I don't see where do we check that scale is valid (e.g. not -1 or 111111111111).
I think it's needed to add to `DataTypeDateTime64` constructor

Reply:
Ok, I see it is in `DataTypeDecimalBase`.

Reply:
It is checked in base class `DataTypeDecimalBase`, and also there are some tests:
https://github.com/ClickHouse/ClickHouse/pull/7170/files/37832c94978419326cfd807defb41d5738f11c00#r354714783